### PR TITLE
Pr vision cam

### PR DESCRIPTION
**Pull request by EliaTarasov**
This pr adds vision cam model to be able to test external vision without internal vision plugin (for example by using ROS).

### CODE REVIEW

File: models/vision_cam/model.sdf [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            <near>0.1</near>
+            <far>100</far>
+          </clip>
+          <noise>
+            <type>gaussian</type>
+            <mean>0.0</mean>
+            <stddev>0.001</stddev>
+          </noise>
+        </camera>
+        <plugin name="svo_plugin" filename="libgazebo_ros_camera.so">
```
Comment (TSC21) [inline models/vision_cam/model.sdf]: `svo_plugin`?

Comment (EliaTarasov) [reply models/vision_cam/model.sdf]: I've tested this model with external vision library which is called svo: https://github.com/uzh-rpg/rpg_svo. But it's up to user what name for plugin is preffered. We may call it vision_plugin.

**Review by TSC21 — APPROVED** [review]

File: models/vision_cam/model.sdf [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            <far>100</far>
+          </clip>
+          <noise>
+            <type>gaussian</type>
+            <mean>0.0</mean>
+            <stddev>0.001</stddev>
+          </noise>
+        </camera>
+        <plugin name="vision_cam_plugin" filename="libgazebo_ros_camera.so">
+	  <alwaysOn>true</alwaysOn>
```
Comment (TSC21) [inline models/vision_cam/model.sdf:48]: The indentation is 2 whitespaces.

**Comment by TSC21** [comment]
@EliaTarasov can you explain and exemplify why is this required and how can it be used? This should also be documented on the PX4 Devguide.

**Comment by EliaTarasov** [comment]
@TSC21 
Explanation:
Currently the way we use vision pose estimation (vision plugin) is: 
1. We get an aircraft's world position and attitude from gazebo.
2. We add white noise and random walk for "vision sensor".
3. Send this to px4.

So here vision works always ok and we can't do much testing. 
On the contrary, optical flow plugin allows us to actually test how real optical flow camera(PX4 Flow or other cam) would work if for example:
1. We attach it to aircraft with different orientation.
2. We attach gyro on that cam with different orientation.
3. We choose ground plane not quite enough textured.
4. We choose camera rate too high or to low.

So overall, optical flow plugin allows us to test some important math stuff behind.
What i want to add is the same capability for vision.

Example:
There is good enough ROS visual odometry library [SVO](https://github.com/uzh-rpg/rpg_svo) and it works well both in sim and real tests.

**Comment by TSC21** [comment]
I see the point. But then I guess it makes sense to actually generalize this for other models by adding a sdf element to select the src for vision estimation (Gazebo itself or an external estimator).

**Comment by EliaTarasov** [comment]
Generalization looks better of course. Does it make sense to implement this sdf element on current commit?

**Comment by TSC21** [comment]
Yes I don't see a problem

**Comment by EliaTarasov** [comment]
I've added <vision_source> tag into gazebo vision plugin: [branch](https://github.com/EliaTarasov/sitl_gazebo/tree/pr-vision_cam_sdf_tag)
Solution looks somewhat awkward since plugin is loaded but not used if <vision_source> is "external".
Is it the way it should be done?

**Comment by TSC21** [comment]
> Solution looks somewhat awkward since plugin is loaded but not used if <vision_source> is "external".

That's the point. If the vision source is external, that means that the vision plugin does not have to start, since there is another vision source producing and publishing data to `gazebo_mavlink_interface`.

**Comment by EliaTarasov** [comment]
> If the vision source is external, that means that the vision plugin does not have to start, since there is another vision source producing and publishing data to gazebo_mavlink_interface.

That's true. 
The only solution that comes to mind is to add one additional external_vision_plugin which would have a tag like `<visionPoseTopic>` and it would somehow handle (maybe just remap) data before sending it to gazebo_mavlink_interface. 
This looks redundant, but logical.
Is it worth to implement?

**Comment by TSC21** [comment]
> The only solution that comes to mind is to add one additional external_vision_plugin which would have a tag like `<visionPoseTopic>` and it would somehow handle (maybe just remap) data before sending it to gazebo_mavlink_interface. 

That makes sense to me. Can you implement and commit? Thanks

**Comment by EliaTarasov** [comment]
@TSC21 I did some research regarding last conversation and found that Gazebo vision topic cannot be easily filled by external vision subsystem's output. Basically we have two options:
1. External vision plugin have to subscribe/advertise some external vision (e.g. Ros) topic.
2. External vision subsystem (e.g. Ros) have to subscribe/advertise some Gazebo topic.

First option's drawback is that plugin's code become directly dependent on Ros which is not good for users want to use gazebo sim without Ros.
Second option's drawback is that user have to add some code relates to Gazebo inside external vision subsystem.

I think these two drawback are too much to be handled in master and now i work on integration of vision library directly into sitl_gazebo code (the way Optical Flow now works).
I'm not sure about three things:
1. Is it acceptable to add some additional code which is large enough.
2. How to handle additional repos for this library automatically during px4 firmware build. 
3. When i finish this branch. 

So for now i would suggest to close this pr as it is not completely solve the original task.

**Comment by TSC21** [comment]
@EliaTarasov additional repos may represent a problem because of the licensing and we actually want to not depend on any kind of other repo. The PR itself should be generic in a way we don't need to rely on any other dependency and, at all. depend on ROS (unless the CMake option `BUILD_ROS_INTERFACE` is set to `ON`).

**Comment by TSC21** [comment]
@EliaTarasov you need to rebase this.

**Comment by EliaTarasov** [comment]
Closed as it was found not suitable for the task.

**Pull request closed without merging**